# Block by Block — PPO portfolio allocation

**Question:** Does a PPO portfolio policy beat equal-weight buy-and-hold on the same
held-out dates, after transaction costs?

The executed cells below contain a fresh GPU training run of **1,000,448 steps**,
followed by evaluation. The policy and its matched training-only VecNormalize
statistics are saved together; test normalization is frozen. Helper Python scripts
execute the notebook and independently replay the saved policy, rather than
substituting for the notebook's training code.

To review without retraining, read the executed outputs and run
`python evaluation/replay_ppo.py` from the repository root. Running every cell
in this notebook trains a new policy and may take substantial time.


## Project context

Block by Block is a financial AI hackathon project built for **AI Odyssey**,
combining forecasting, RL trading experiments, quantitative tools and a web UI.
The exact edition year has not yet been confirmed; no placement is claimed here.

This is a historical research simulation, not live trading. The measured run
below is new evaluation work and is not presented as a hackathon award result.


# Imports

Load essential libraries for data processing, visualization, reinforcement learning, and explainability.

In [2]:
from pathlib import Path
import os, sys, json
import warnings

def format_warning_for_report(message, category, filename, lineno, line=None):
    """Keep warning content while omitting machine-specific directories."""
    return f'{Path(filename).name}:{lineno}: {category.__name__}: {message}\n'

warnings.formatwarning = format_warning_for_report

def locate_experiment_root():
    for base in (Path.cwd().resolve(), *Path.cwd().resolve().parents):
        candidate = base / 'evaluation'
        if (candidate / 'data-sources').is_dir() and (candidate / 'run-assets').is_dir():
            return candidate
    raise FileNotFoundError('Run from this repository with its evaluation inputs present.')

EXPERIMENT_ROOT = locate_experiment_root()
sys.path.insert(0, str(EXPERIMENT_ROOT / 'notebooks'))
if not (EXPERIMENT_ROOT / 'run-assets').exists():
    raise FileNotFoundError('The evaluation input snapshot must be present.')
os.environ['CUDA_VISIBLE_DEVICES'] = '0'
sys.path.insert(0, str(EXPERIMENT_ROOT / 'runtime_vendor'))
PROJECT_PATH = str(EXPERIMENT_ROOT / 'runs' / 'ppo')
Path(PROJECT_PATH, 'checkpoints').mkdir(parents=True, exist_ok=True)
Path(PROJECT_PATH, 'plots').mkdir(parents=True, exist_ok=True)
print('Run outputs:', Path('runs') / 'ppo')


Run outputs: runs/ppo


In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import torch
import shap
from IPython.display import display
from datetime import datetime
from finrl.meta.preprocessor.preprocessors import FeatureEngineer, data_split
from finrl.meta.env_stock_trading.env_stocktrading import StockTradingEnv
from stable_baselines3 import PPO
from stable_baselines3.common.vec_env import DummyVecEnv, VecNormalize
from stable_baselines3.common.callbacks import BaseCallback
if not torch.cuda.is_available():
    raise RuntimeError('A PyTorch CUDA GPU is required for this run.')
torch.set_num_threads(4)
print('PyTorch:', torch.__version__, 'GPU:', torch.cuda.get_device_name(0))


auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


PyTorch: 2.14.0+cu130 GPU: NVIDIA RTX A3000 12GB Laptop GPU


# Configuration

Define trading parameters including date ranges, ticker symbols, technical indicators, and portfolio settings.

In [4]:
TRAIN_START_DATE = '2016-01-01'
TRAIN_END_DATE = '2023-01-01'
TEST_START_DATE = '2023-01-01'
TEST_END_DATE = '2025-01-01'

TICKER_LIST = [
    # Green Energy
    "NEE", "FSLR", "BEP", "PLUG",

    # Oil & Gas
    "XOM", "CVX", "COP", "SLB",

    # Banking & Finance
    "JPM", "BAC", "WFC", "GS", "MS",

    # Utilities / General Energy
    "DUK", "SO", "PCG", "EIX",

    # IT / Tech
    "AAPL", "MSFT", "GOOGL", "NVDA", "AMZN", "META", "ORCL", "IBM"
]

INDICATORS_LIST = ['macd', 'rsi_30', 'cci_30', 'dx_30', 'close_30_sma', 'close_60_sma']

INITIAL_AMOUNT = 100000
TRANSACTION_COST_PCT = 0.001
REWARD_SCALING = 1e-4

# Load pinned historical responses
Read real cached Yahoo OHLCV. Source URLs and checksums are in the data manifest.

In [6]:
from datetime import timezone
records = []
for ticker in TICKER_LIST:
    raw = json.loads((EXPERIMENT_ROOT / 'run-assets' / 'ppo-history' / (ticker + '.json')).read_text())
    result = raw['chart']['result'][0]
    quote = result['indicators']['quote'][0]
    adjusted = result['indicators']['adjclose'][0]['adjclose']
    for i, stamp in enumerate(result['timestamp']):
        records.append({'date': datetime.fromtimestamp(stamp, tz=timezone.utc).strftime('%Y-%m-%d'),
                        'tic': ticker, 'open': quote['open'][i], 'high': quote['high'][i],
                        'low': quote['low'][i], 'close': quote['close'][i],
                        'adj_close': adjusted[i], 'volume': quote['volume'][i]})
df = pd.DataFrame(records)
print('Raw rows:', len(df), 'tickers:', df.tic.nunique())
print('Missing values before cleaning:')
display(df.isna().sum().to_frame('missing'))


Raw rows: 56600 tickers: 25
Missing values before cleaning:


,missing
date,0
tic,0
open,0
high,0
low,0
close,0
adj_close,0
volume,0


# Inspect and clean input rows
Report missing values, keep positive prices, and sort actual date/ticker observations.

In [7]:
df_finrl = df.dropna().copy()
df_finrl = df_finrl[df_finrl['close'] > 0].sort_values(['date', 'tic']).reset_index(drop=True)
print('Valid input rows:', len(df_finrl))


Valid input rows: 56600


# Feature Engineering

Generate technical indicators (MACD, RSI, CCI, DX, SMA) to enrich the state space for the RL agent.

In [8]:
fe = FeatureEngineer(use_technical_indicator=True, tech_indicator_list=INDICATORS_LIST,
                     use_vix=False, use_turbulence=False, user_defined_feature=False)
# Avoid the upstream preprocessing method's backward fill of missing indicators.
# Use causal indicators, discard nonfinite warm-up rows, and retain complete dates.
processed_df = fe.add_technical_indicator(fe.clean_data(df_finrl))
processed_df = processed_df.replace([np.inf, -np.inf], np.nan).dropna()
expected_stocks = processed_df.tic.nunique()
counts = processed_df.groupby('date')['tic'].nunique()
complete_dates = counts[counts == expected_stocks].index
processed_df = processed_df[processed_df.date.isin(complete_dates)].sort_values(['date', 'tic'])
print('Processed rows:', len(processed_df), 'tickers:', expected_stocks)
print('Excluded configured tickers:', sorted(set(TICKER_LIST) - set(processed_df.tic)))
display(processed_df.head())


Processed rows: 56600 tickers: 25
Excluded configured tickers: []


,date,tic,open,high,low,close,adj_close,volume,macd,rsi_30,cci_30,dx_30,close_30_sma,close_60_sma
0,2016-01-04,AAPL,25.652500,26.342501,25.500000,26.337500,23.688671,270597600,0.0,50.0,0.0,0.0,26.337500,26.337500
2264,2016-01-04,AMZN,32.814499,32.886002,31.375500,31.849501,31.849501,186290000,0.0,50.0,0.0,0.0,31.849501,31.849501
4528,2016-01-04,BAC,16.450001,16.490000,16.250000,16.430000,12.988195,114888000,0.0,50.0,0.0,0.0,16.430000,16.430000
6792,2016-01-04,BEP,13.930189,13.930189,13.471889,13.562483,7.942327,60048,0.0,50.0,0.0,0.0,13.562483,13.562483
9056,2016-01-04,COP,46.410000,47.770000,46.080002,46.880001,33.748306,8918000,0.0,50.0,0.0,0.0,46.880001,46.880001


# Chronological train/test split
Training dates come from TRAIN_START_DATE/TRAIN_END_DATE; test dates from TEST_START_DATE/TEST_END_DATE. All model fitting and observation normalization use training data only.

In [9]:
train_df = data_split(processed_df, TRAIN_START_DATE, TRAIN_END_DATE)
test_df = data_split(processed_df, TEST_START_DATE, TEST_END_DATE)

print(f"Training: {len(train_df)} samples ({TRAIN_START_DATE} to {TRAIN_END_DATE})")
print(f"Testing:  {len(test_df)} samples ({TEST_START_DATE} to {TEST_END_DATE})")

Training: 44050 samples (2016-01-01 to 2023-01-01)
Testing:  12550 samples (2023-01-01 to 2025-01-01)


# Create Training Environment

Initialize the stock trading environment with the configured state space and action space.

In [10]:
stock_dimension = len(train_df.tic.unique())
state_space = 1 + 2*stock_dimension + len(INDICATORS_LIST)*stock_dimension

env_kwargs = {
    "hmax": 100,
    "initial_amount": INITIAL_AMOUNT,
    "num_stock_shares": [0] * stock_dimension,
    "buy_cost_pct": [TRANSACTION_COST_PCT] * stock_dimension,
    "sell_cost_pct": [TRANSACTION_COST_PCT] * stock_dimension,
    "state_space": state_space,
    "stock_dim": stock_dimension,
    "tech_indicator_list": INDICATORS_LIST,
    "action_space": stock_dimension,
    "reward_scaling": REWARD_SCALING
}

train_env = DummyVecEnv([lambda: StockTradingEnv(df=train_df, **env_kwargs)])
train_env = VecNormalize(train_env, norm_obs=True, norm_reward=True, gamma=0.99)
print(f"Stocks: {stock_dimension} | State space: {state_space} | Action space: {stock_dimension}")

Stocks: 25 | State space: 201 | Action space: 25


# Train PPO
Fit a PPO policy and save its matching normalization statistics. Training is experimental; no optimality or outperformance is assumed.

In [11]:
import time
class ProgressCheckpoint(BaseCallback):
    def _on_training_start(self):
        self.started = time.monotonic()
    def _on_step(self):
        if self.num_timesteps % 10000 == 0:
            elapsed = time.monotonic() - self.started
            Path(PROJECT_PATH, 'training_progress.json').write_text(json.dumps(
                {'timesteps': self.num_timesteps, 'elapsed_seconds': elapsed,
                 'steps_per_second': self.num_timesteps / elapsed}, indent=2))
        return True

TOTAL_TIMESTEPS = int(os.environ.get('PPO_TIMESTEPS', '1000000'))
model = PPO('MlpPolicy', train_env, learning_rate=0.00003, n_steps=512,
            batch_size=256, n_epochs=10, gamma=0.99, ent_coef=0.005,
            gae_lambda=0.95, clip_range=0.2, seed=42, device='cuda', verbose=0)
assert model.device.type == 'cuda'
print('Training a fresh PPO policy on', model.device, 'for', TOTAL_TIMESTEPS, 'steps.')
model.learn(total_timesteps=TOTAL_TIMESTEPS, callback=ProgressCheckpoint())
timestamp = datetime.now().strftime('%Y%m%d_%H%M%S')
checkpoint_path = f'{PROJECT_PATH}/checkpoints/ppo_model_trained_{timestamp}'
model.save(checkpoint_path)
normalization_path = checkpoint_path + '_vecnormalize.pkl'
train_env.save(normalization_path)
print('Actual training steps:', model.num_timesteps, 'saved matching model and normalizer.')


on_policy_algorithm.py:150: UserWarning: You are trying to run PPO on the GPU, but it is primarily intended to run on the CPU when not using a CNN policy (you are using ActorCriticPolicy which should be a MlpPolicy). See https://github.com/DLR-RM/stable-baselines3/issues/1245 for more info. You can pass `device='cpu'` or `export CUDA_VISIBLE_DEVICES=` to force using the CPU.Note: The model will train, but the GPU utilization will be poor and the training might take longer than on CPU.
  warnings.warn(


Training a fresh PPO policy on cuda for 1000000 steps.


day: 1761, episode: 10
begin_total_asset: 100000.00
end_total_asset: 135799.60
total_reward: 35799.60
total_cost: 84009.48
total_trades: 34010
Sharpe: 0.310


day: 1761, episode: 20
begin_total_asset: 100000.00
end_total_asset: 119061.48
total_reward: 19061.48
total_cost: 78459.45
total_trades: 33794
Sharpe: 0.224


day: 1761, episode: 30
begin_total_asset: 100000.00
end_total_asset: 102618.24
total_reward: 2618.24
total_cost: 70496.39
total_trades: 33106
Sharpe: 0.127


day: 1761, episode: 40
begin_total_asset: 100000.00
end_total_asset: 106159.82
total_reward: 6159.82
total_cost: 67807.71
total_trades: 32969
Sharpe: 0.149


day: 1761, episode: 50
begin_total_asset: 100000.00
end_total_asset: 203993.15
total_reward: 103993.15
total_cost: 83329.46
total_trades: 33989
Sharpe: 0.571


day: 1761, episode: 60
begin_total_asset: 100000.00
end_total_asset: 117902.18
total_reward: 17902.18
total_cost: 60575.25
total_trades: 31880
Sharpe: 0.218


day: 1761, episode: 70
begin_total_asset: 100000.00
end_total_asset: 245242.12
total_reward: 145242.12
total_cost: 67381.26
total_trades: 32827
Sharpe: 0.654


day: 1761, episode: 80
begin_total_asset: 100000.00
end_total_asset: 259715.11
total_reward: 159715.11
total_cost: 48530.59
total_trades: 30437
Sharpe: 0.654


day: 1761, episode: 90
begin_total_asset: 100000.00
end_total_asset: 496285.01
total_reward: 396285.01
total_cost: 36006.31
total_trades: 28999
Sharpe: 1.039


day: 1761, episode: 100
begin_total_asset: 100000.00
end_total_asset: 559571.73
total_reward: 459571.73
total_cost: 30615.77
total_trades: 28095
Sharpe: 1.056


day: 1761, episode: 110
begin_total_asset: 100000.00
end_total_asset: 647514.38
total_reward: 547514.38
total_cost: 25925.08
total_trades: 27591
Sharpe: 1.171


day: 1761, episode: 120
begin_total_asset: 100000.00
end_total_asset: 780642.36
total_reward: 680642.36
total_cost: 25153.85
total_trades: 26987
Sharpe: 1.248


day: 1761, episode: 130
begin_total_asset: 100000.00
end_total_asset: 673612.85
total_reward: 573612.85
total_cost: 18325.13
total_trades: 26251
Sharpe: 1.155


day: 1761, episode: 140
begin_total_asset: 100000.00
end_total_asset: 732569.88
total_reward: 632569.88
total_cost: 17425.14
total_trades: 26238
Sharpe: 1.199


day: 1761, episode: 150
begin_total_asset: 100000.00
end_total_asset: 764620.73
total_reward: 664620.73
total_cost: 14531.09
total_trades: 25665
Sharpe: 1.199


day: 1761, episode: 160
begin_total_asset: 100000.00
end_total_asset: 751725.05
total_reward: 651725.05
total_cost: 10755.52
total_trades: 25169
Sharpe: 1.173


day: 1761, episode: 170
begin_total_asset: 100000.00
end_total_asset: 691446.88
total_reward: 591446.88
total_cost: 10061.40
total_trades: 24951
Sharpe: 1.132


day: 1761, episode: 180
begin_total_asset: 100000.00
end_total_asset: 742885.54
total_reward: 642885.54
total_cost: 10077.92
total_trades: 24967
Sharpe: 1.160


day: 1761, episode: 190
begin_total_asset: 100000.00
end_total_asset: 833411.56
total_reward: 733411.56
total_cost: 11007.57
total_trades: 24742
Sharpe: 1.165


day: 1761, episode: 200
begin_total_asset: 100000.00
end_total_asset: 733094.22
total_reward: 633094.22
total_cost: 8684.21
total_trades: 24614
Sharpe: 1.147


day: 1761, episode: 210
begin_total_asset: 100000.00
end_total_asset: 833099.32
total_reward: 733099.32
total_cost: 10026.74
total_trades: 24606
Sharpe: 1.149


day: 1761, episode: 220
begin_total_asset: 100000.00
end_total_asset: 771994.74
total_reward: 671994.74
total_cost: 11889.93
total_trades: 24984
Sharpe: 1.067


day: 1761, episode: 230
begin_total_asset: 100000.00
end_total_asset: 551156.55
total_reward: 451156.55
total_cost: 12746.28
total_trades: 25004
Sharpe: 0.852


day: 1761, episode: 240
begin_total_asset: 100000.00
end_total_asset: 777288.99
total_reward: 677288.99
total_cost: 14206.45
total_trades: 25090
Sharpe: 1.072


day: 1761, episode: 250
begin_total_asset: 100000.00
end_total_asset: 925411.58
total_reward: 825411.58
total_cost: 13597.13
total_trades: 25264
Sharpe: 1.166


day: 1761, episode: 260
begin_total_asset: 100000.00
end_total_asset: 817985.99
total_reward: 717985.99
total_cost: 14723.52
total_trades: 25127
Sharpe: 1.093


day: 1761, episode: 270
begin_total_asset: 100000.00
end_total_asset: 805382.11
total_reward: 705382.11
total_cost: 13693.96
total_trades: 25328
Sharpe: 1.068


day: 1761, episode: 280
begin_total_asset: 100000.00
end_total_asset: 760850.82
total_reward: 660850.82
total_cost: 14143.40
total_trades: 25180
Sharpe: 1.019


day: 1761, episode: 290
begin_total_asset: 100000.00
end_total_asset: 554708.69
total_reward: 454708.69
total_cost: 15969.68
total_trades: 25568
Sharpe: 0.840


day: 1761, episode: 300
begin_total_asset: 100000.00
end_total_asset: 782749.12
total_reward: 682749.12
total_cost: 16816.21
total_trades: 25698
Sharpe: 1.067


day: 1761, episode: 310
begin_total_asset: 100000.00
end_total_asset: 739169.44
total_reward: 639169.44
total_cost: 15137.34
total_trades: 25432
Sharpe: 1.028


day: 1761, episode: 320
begin_total_asset: 100000.00
end_total_asset: 782068.81
total_reward: 682068.81
total_cost: 18250.70
total_trades: 25847
Sharpe: 1.097


day: 1761, episode: 330
begin_total_asset: 100000.00
end_total_asset: 759274.69
total_reward: 659274.69
total_cost: 15593.02
total_trades: 25502
Sharpe: 1.053


day: 1761, episode: 340
begin_total_asset: 100000.00
end_total_asset: 811043.96
total_reward: 711043.96
total_cost: 17769.33
total_trades: 25916
Sharpe: 1.163


day: 1761, episode: 350
begin_total_asset: 100000.00
end_total_asset: 710926.79
total_reward: 610926.79
total_cost: 15333.09
total_trades: 25502
Sharpe: 1.015


day: 1761, episode: 360
begin_total_asset: 100000.00
end_total_asset: 923105.28
total_reward: 823105.28
total_cost: 14706.13
total_trades: 25198
Sharpe: 1.152


day: 1761, episode: 370
begin_total_asset: 100000.00
end_total_asset: 927773.40
total_reward: 827773.40
total_cost: 13065.42
total_trades: 25046
Sharpe: 1.207


day: 1761, episode: 380
begin_total_asset: 100000.00
end_total_asset: 983420.12
total_reward: 883420.12
total_cost: 18902.56
total_trades: 25813
Sharpe: 1.183


day: 1761, episode: 390
begin_total_asset: 100000.00
end_total_asset: 854932.56
total_reward: 754932.56
total_cost: 17163.88
total_trades: 25355
Sharpe: 1.130


day: 1761, episode: 400
begin_total_asset: 100000.00
end_total_asset: 708468.87
total_reward: 608468.87
total_cost: 12707.36
total_trades: 24648
Sharpe: 0.988


day: 1761, episode: 410
begin_total_asset: 100000.00
end_total_asset: 922561.39
total_reward: 822561.39
total_cost: 16343.03
total_trades: 25498
Sharpe: 1.222


day: 1761, episode: 420
begin_total_asset: 100000.00
end_total_asset: 965965.23
total_reward: 865965.23
total_cost: 15845.95
total_trades: 25260
Sharpe: 1.195


day: 1761, episode: 430
begin_total_asset: 100000.00
end_total_asset: 1059386.24
total_reward: 959386.24
total_cost: 13701.21
total_trades: 25030
Sharpe: 1.255


day: 1761, episode: 440
begin_total_asset: 100000.00
end_total_asset: 977201.06
total_reward: 877201.06
total_cost: 13084.33
total_trades: 24827
Sharpe: 1.176


day: 1761, episode: 450
begin_total_asset: 100000.00
end_total_asset: 991073.33
total_reward: 891073.33
total_cost: 16694.36
total_trades: 25203
Sharpe: 1.177


day: 1761, episode: 460
begin_total_asset: 100000.00
end_total_asset: 1104104.69
total_reward: 1004104.69
total_cost: 19131.77
total_trades: 25700
Sharpe: 1.290


day: 1761, episode: 470
begin_total_asset: 100000.00
end_total_asset: 1123825.23
total_reward: 1023825.23
total_cost: 20220.50
total_trades: 25802
Sharpe: 1.326


day: 1761, episode: 480
begin_total_asset: 100000.00
end_total_asset: 1084495.01
total_reward: 984495.01
total_cost: 20624.80
total_trades: 26017
Sharpe: 1.315


day: 1761, episode: 490
begin_total_asset: 100000.00
end_total_asset: 1135204.57
total_reward: 1035204.57
total_cost: 17724.99
total_trades: 25507
Sharpe: 1.338


day: 1761, episode: 500
begin_total_asset: 100000.00
end_total_asset: 1186402.09
total_reward: 1086402.09
total_cost: 19121.63
total_trades: 25872
Sharpe: 1.341


day: 1761, episode: 510
begin_total_asset: 100000.00
end_total_asset: 1467975.42
total_reward: 1367975.42
total_cost: 19393.87
total_trades: 25688
Sharpe: 1.394


day: 1761, episode: 520
begin_total_asset: 100000.00
end_total_asset: 1294360.82
total_reward: 1194360.82
total_cost: 19780.03
total_trades: 25684
Sharpe: 1.323


day: 1761, episode: 530
begin_total_asset: 100000.00
end_total_asset: 1348625.82
total_reward: 1248625.82
total_cost: 12835.82
total_trades: 24735
Sharpe: 1.429


day: 1761, episode: 540
begin_total_asset: 100000.00
end_total_asset: 1215547.73
total_reward: 1115547.73
total_cost: 14299.31
total_trades: 24915
Sharpe: 1.363


day: 1761, episode: 550
begin_total_asset: 100000.00
end_total_asset: 1188288.84
total_reward: 1088288.84
total_cost: 14476.22
total_trades: 24889
Sharpe: 1.357


day: 1761, episode: 560
begin_total_asset: 100000.00
end_total_asset: 1190070.67
total_reward: 1090070.67
total_cost: 14894.72
total_trades: 24920
Sharpe: 1.307


Actual training steps: 1000448 saved matching model and normalizer.


# Backtest Model

Evaluate the trained agent on unseen test data and calculate performance metrics.

In [12]:
# Feature and ticker ordering must match the trained policy.
assert list(train_df.tic.unique()) == list(test_df.tic.unique()), (
    "Train/test ticker ordering differs: align data before evaluation."
)
stock_dimension = len(test_df.tic.unique())
state_space = 1 + 2*stock_dimension + len(INDICATORS_LIST)*stock_dimension

env_kwargs = {
    "hmax": 100,
    "initial_amount": INITIAL_AMOUNT,
    "num_stock_shares": [0] * stock_dimension,
    "buy_cost_pct": [TRANSACTION_COST_PCT] * stock_dimension,
    "sell_cost_pct": [TRANSACTION_COST_PCT] * stock_dimension,
    "state_space": state_space,
    "stock_dim": stock_dimension,
    "tech_indicator_list": INDICATORS_LIST,
    "action_space": stock_dimension,
    "reward_scaling": REWARD_SCALING
}

test_env = StockTradingEnv(df=test_df, **env_kwargs)
# Restore TRAINING statistics; never fit a normalizer on held-out observations.
# We step the raw environment to retain raw states for cash/position plots,
# and explicitly normalize every policy input with this frozen normalizer.
test_vec_env = DummyVecEnv([lambda: test_env])
eval_normalizer = VecNormalize.load(normalization_path, test_vec_env)
eval_normalizer.training = False
eval_normalizer.norm_reward = False
test_obs, info = test_env.reset()

actions_taken, states_log = [], []

done = False
while not done:
    normalized_obs = eval_normalizer.normalize_obs(
        np.asarray(test_obs, dtype=np.float32).reshape(1, -1)
    )
    action_batch, _states = model.predict(normalized_obs, deterministic=True)
    action = action_batch[0]
    step_result = test_env.step(action)

    if len(step_result) == 5:
        test_obs, rewards, terminated, truncated, info = step_result
        done = terminated or truncated
    else:
        test_obs, rewards, done, info = step_result

    actions_taken.append(action)
    states_log.append(test_obs)

# Use canonical dated asset memory: terminal step can repeat the last value.
account_frame = test_env.save_asset_memory().copy()
account_values = account_frame['account_value'].to_numpy()
print(f"Backtest observations: {len(account_frame)}; dates: "
      f"{account_frame['date'].iloc[0]} to {account_frame['date'].iloc[-1]}")

Backtest observations: 502; dates: 2023-01-03 to 2024-12-31


# Does the policy beat a simple benchmark, and at what risk?
Keep one return/drawdown figure and a metrics table. Both strategies use exactly the same dates and stocks. The equal-weight benchmark buys fractional shares once, pays the configured entry fee, and is then marked to market without liquidation; PPO pays its environment transaction costs. Sharpe uses daily observations, 252 periods/year, and a zero risk-free rate. These assumptions are approximations, not trading advice or evidence of deployment.

In [13]:
from notebook_figures import trading_comparison, trading_metrics, plot_trading
from notebook_figures import finrl_feature_names, plot_shap_action

price_frame = test_df.pivot(index='date', columns='tic', values='close')
comparison = trading_comparison(account_frame, price_frame, INITIAL_AMOUNT, TRANSACTION_COST_PCT)
metrics = trading_metrics(comparison, INITIAL_AMOUNT)
display(metrics.round(3))
fig = plot_trading(comparison, INITIAL_AMOUNT, f'{PROJECT_PATH}/plots/benchmark_and_drawdown.png')
plt.show()
plt.close(fig)
comparison.to_csv(f'{PROJECT_PATH}/dated_backtest.csv', index_label='date')
metrics.to_csv(f'{PROJECT_PATH}/backtest_comparison_metrics.csv')

,return_pct,max_drawdown_pct,annualized_volatility_pct,sharpe_rf_zero
strategy,,,,
PPO,53.146,-8.267,14.888,1.515
Equal-weight hold,80.781,-13.468,16.386,1.900


902589571.py:9: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


# Optional: which inputs influence ONE stock action?
This explains the trained policy's action for a selected ticker on sampled test states. It does not explain returns, prove causal importance, or summarize all stock actions. FinRL state order is cash, prices, holdings, then indicator-major blocks. Do not reshape or average signed SHAP values across unrelated outputs to make a plot fit. Disabled by default because Kernel SHAP is expensive. Inspect this diagnostic after the primary comparison.

In [14]:
RUN_SHAP = False
SHAP_TICKER = 'AAPL'
if RUN_SHAP:
    import pickle
    tickers = test_df.loc[test_df.index == test_df.index.min(), 'tic'].tolist()
    feature_names = finrl_feature_names(tickers, INDICATORS_LIST)
    states_array = np.asarray(states_log)
    assert states_array.ndim == 2 and states_array.shape[1] == len(feature_names)
    assert SHAP_TICKER in tickers and len(states_array) >= 12
    action_index = tickers.index(SHAP_TICKER)
    background_count = min(30, len(states_array) // 2)
    background_data = states_array[:background_count]
    test_data = states_array[background_count:background_count + 20]
    def predict_selected_action(states):
        actions, _ = model.predict(
            eval_normalizer.normalize_obs(np.asarray(states, dtype=np.float32)),
            deterministic=True,
        )
        return np.asarray(actions)[:, action_index]
    explainer = shap.KernelExplainer(predict_selected_action, background_data)
    shap_values = np.asarray(explainer.shap_values(test_data, nsamples=200))
    if shap_values.shape != test_data.shape:
        raise ValueError('Unexpected scalar-action SHAP dimensions; inspect the library API.')
    fig = plot_shap_action(shap_values, feature_names, SHAP_TICKER,
                           f'{PROJECT_PATH}/plots/selected_action_shap.png')
    plt.show()
    plt.close(fig)
    with open(f'{PROJECT_PATH}/checkpoints/selected_action_shap.pkl', 'wb') as handle:
        pickle.dump({'ticker': SHAP_TICKER, 'feature_names': feature_names,
                     'test_data': test_data, 'shap_values': shap_values}, handle)
else:
    print('Optional per-stock SHAP diagnostic skipped.')

Optional per-stock SHAP diagnostic skipped.


# Findings and limitations
The final interpretation below reports the completed comparison. This is one
seeded simulation with a selected stock universe. Multiple-seed stability,
survivorship bias, live execution/slippage, and out-of-period robustness remain
unverified. Optional SHAP is disabled and is not presented as executed evidence.


# Save the model and its matching normalizer

In [15]:
model.save(f'{PROJECT_PATH}/trained_ppo_trading_model')
train_env.save(f'{PROJECT_PATH}/trained_ppo_vecnormalize.pkl')
np.save(f'{PROJECT_PATH}/checkpoints/actions_taken.npy', np.asarray(actions_taken))
print('Saved: model, training normalizer, dated_backtest.csv, comparison metrics, and the comparison figure.')

Saved: model, training normalizer, dated_backtest.csv, comparison metrics, and the comparison figure.


## Measured findings from this run
The next cell compares this fresh, seeded GPU run with the baseline. Training timesteps, data coverage, dates, software, and device are recorded. The observed result is one historical simulation; it does not establish robustness across seeds, future periods, or deployment. Price observations use the downloaded Yahoo Close series; dividends are not separately credited. Review survivorship and feature timing before stronger conclusions.

In [16]:
summary = {
    'mode': 'fresh_gpu_training', 'configured_timesteps': TOTAL_TIMESTEPS,
    'actual_timesteps': model.num_timesteps, 'seed': 42,
    'train_rows': len(train_df), 'test_rows': len(test_df),
    'tickers': list(train_df.tic.unique()),
    'train_start': train_df.date.min(), 'train_end': train_df.date.max(),
    'test_start': test_df.date.min(), 'test_end': test_df.date.max(),
    'torch_version': torch.__version__, 'gpu': torch.cuda.get_device_name(0),
    'metrics': metrics.to_dict(orient='index'),
}
Path(PROJECT_PATH, 'run_summary.json').write_text(json.dumps(summary, indent=2) + '\n')
ppo_return = metrics.loc['PPO', 'return_pct']
hold_return = metrics.loc['Equal-weight hold', 'return_pct']
print(f'PPO return: {ppo_return:+.2f}%; equal-weight hold: {hold_return:+.2f}%.')
print(f'Difference: {ppo_return - hold_return:+.2f} percentage points.')
print('This is one seeded simulation, not an estimate of live investment returns.')

PPO return: +53.15%; equal-weight hold: +80.78%.
Difference: -27.64 percentage points.
This is one seeded simulation, not an estimate of live investment returns.


## Interpretation of the completed run (2026-10-06 23:27 UTC)

**Run:** seed 42, **1,000,448 actual PPO training steps**
on CUDA, across **25 stocks**. Training observations span
**2016-01-04–2022-12-30**; the held-out backtest
spans **2023-01-03–2024-12-31**. The complete
dated account series, fees, input audit, and software/device metadata are saved.

| Strategy | Return (%) | Max drawdown (%) | Sharpe (zero risk-free) |
|---|---:|---:|---:|
| PPO | +53.15 | -8.27 | 1.515 |
| Equal-weight hold | +80.78 | -13.47 | 1.900 |
| PPO: one-observation delay | +53.64 | -8.25 | 1.529 |


**Observed outcome:** PPO trails equal-weight hold by
**-27.64 percentage points** in this period. Compare
drawdown and volatility as well as total return: a strategy retaining cash can
have a different risk profile from a fully invested benchmark. This is not a
risk-matched comparison or evidence of stable profitability.

**Execution sensitivity:** the original FinRL simulation observes close-t prices
and indicators and trades at that same close. This is an optimistic execution
convention. A separate replay chooses an action using close t, then executes the
fixed action at the next observation's close; its first order is zero. Delayed
PPO return is **+53.64%**, a change of
**+0.50 points** from the original backtest. The
precommitted benchmark still buys at the first date. This audit changes execution
at evaluation; the policy was not retrained with delay, so this is sensitivity
evidence, not a new optimized trading model.

**Order behavior:** the delayed replay has **49 nonzero filled stock/date orders** (35 buys, 14 sells) across **15 dates**, with **$117.01** in fees. The upstream trade counter is **6,594**, but includes attempts with zero fill. Use executed action memory to assess trading activity; frequent requests do not establish frequent executed rebalancing or adaptive trading skill.

**What was verified:** reloading the saved policy with its matching frozen
VecNormalize statistics reproduces the original dated held-out account values.
Appending future rows leaves earlier technical indicators unchanged at two
audited cutoffs (2019 and 2023). The raw snapshot has no missing OHLCV values;
short initial indicator windows follow upstream stockstats behavior.
The upstream environment's tradability guard compares the first indicator with
`True`; no MACD values trigger that guard in this snapshot (see data_audit.json).
This assumption must be revisited if the data or feature schema changes.

**Limits and next experiment:** the stock universe is selected rather than a
verified point-in-time universe; dividends are not credited separately, and
slippage is absent. Develop a policy using delayed execution and multiple seeds,
select settings on development periods, and confirm on an untouched later period.
Do not tune this policy against the now-inspected 2023–2024 comparison. Optional
single-action SHAP remains unrun and does not explain portfolio returns.

Reproduce the saved-policy replay and delay audit without training again:
`python evaluation/replay_ppo.py` from the workspace root. The replay
outputs are separate under `runs/ppo-replay/`.
